src/software/ai/hl/stp/tactic/tactic_visitor.h

```diff
@@ -42,22 +42,22 @@ class TacticVisitor
      * @param tactic The tactic to visit
      */
 
-    virtual void visit(const CherryPickTactic &tactic)         = 0;
-    virtual void visit(const ShadowFreekickerTactic &tactic) = 0;
-    virtual void visit(const GoalieTactic &tactic) = 0;
-    virtual void visit(const CreaseDefenderTactic &tactic) = 0;
-    virtual void visit(const ShadowEnemyTactic &tactic) = 0;
-    virtual void visit(const MoveTactic &tactic) = 0;
-    virtual void visit(const ChipTactic &tactic) = 0;
-    virtual void visit(const KickoffChipTactic &tactic)        = 0;
-    virtual void visit(const StopTactic &tactic) = 0;
-    virtual void visit(const PatrolTactic &tactic)             = 0;
-    virtual void visit(const PenaltyKickTactic &tactic) = 0;
-    virtual void visit(const PenaltySetupTactic &tactic)       = 0;
-    virtual void visit(const ReceiverTactic &tactic) = 0;
-    virtual void visit(const ShootGoalTactic &tactic) = 0;
-    virtual void visit(const PasserTactic &tactic) = 0;
-    virtual void visit(const MoveTestTactic &tactic) = 0;
-    virtual void visit(const StopTestTactic &tactic) = 0;
-    virtual void visit(const DefenseShadowEnemyTactic &tactic) = 0;
+    virtual void visit(CherryPickTactic &tactic)         = 0;
```

Comment:
We may want to consider having two different versions of the generic visitor, since in most cases we _do_ want to keep the guarantee of `const`.

@garethellis0 thoughts? I suppose we need to maintain two versions? I don't think we can use any `const` casting since these aren't pointers.

Reply:
What would the class hierarchy look like for two versions? Would it have to be an empty TacticVisitor class with an Immutable and Mutable child classes?

Reply:
Yeah, I don't think there would be any benefit to having a class hierarchy here. We'll need to have two versions of the `TacticVisitor`, one with `virtual void visit(const SomeTactic& tactic)` and one with `virtual void visit(SomeTactic& tactic)`. I don't think there's any way to unify those function from a class-perspective, as far as the C++ type system is concerned they're entirely different functions.

Reply:
oh i see, i was thinking I had to for the accept function but i guess i can just overload it

Reply:
I'd do `NonMutableTacticVisitor`  and `MutableTacticVisitor`, as there's no real default assumption of whether something is mutable or not.